src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -27,8 +25,50 @@ public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextCh
             Debug.Assert(!changeRanges.IsDefault);
 
             _newText = newText;
-            _weakOldText = new WeakReference<SourceText>(oldText);
-            _changes = changeRanges;
+            _info = new ChangeInfo(changeRanges, new WeakReference<SourceText>(oldText), (oldText as ChangedText)?._info);
+        }
+
+        private class ChangeInfo
+        {
+            public ImmutableArray<TextChangeRange> ChangeRanges { get; }
+
+            // store old text weakly so we don't form unwanted chains of old texts (especially chains of ChangedTexts)
+            // used to identify the changes in GetChangeRanges.
+            public WeakReference<SourceText> WeakOldText { get; }
+
+            public ChangeInfo Previous { get; private set; }
+
+            public ChangeInfo(ImmutableArray<TextChangeRange> changeRanges, WeakReference<SourceText> weakOldText, ChangeInfo previous)
+            {
+                this.ChangeRanges = changeRanges;
+                this.WeakOldText = weakOldText;
+                this.Previous = previous;
+                Clean();
+            }
+
+            // clean up 
+            private void Clean()
+            {
+                // look for last info in the chain that still has reference to old text
+                ChangeInfo lastInfo = this;
+                for (var info = this; info != null; info = info.Previous)
```

Comment:
Perhaps limit the length of the preserved chain to some "reasonable" number (100? 1000?).

That is just in case if one text get stuck in Gen2 heap for minutes or GC gets lazy on a big machine.
It seems that at some chain length it might be counterproductive to merge changes.